go/kvstore/cache.go

```diff
@@ -42,45 +44,104 @@ func (e KVRevisionCacheError) Error() string {
 	return e.Message
 }
 
-func (k *KVRevisionCache) PutCheck(mctx libkb.MetaContext, entryID keybase1.KVEntryID, entryHash string, teamKeyGen keybase1.PerTeamKeyGeneration, revision int) (err error) {
+// Hash is a sha256 on the input string. If the string is empty, then Hash will also be an
+// empty string for tracking deleted entries in perpetuity.
+func (k *KVRevisionCache) hash(ciphertext *string) string {
+	if ciphertext == nil || len(*ciphertext) == 0 || *ciphertext == DELETED_OR_NONEXISTENT {
+		return DELETED_OR_NONEXISTENT
+	}
+	b := sha256.Sum256([]byte(*ciphertext))
+	return hex.EncodeToString(b[:])
+}
+
+func (k *KVRevisionCache) Check(mctx libkb.MetaContext, entryID keybase1.KVEntryID, ciphertext *string, teamKeyGen keybase1.PerTeamKeyGeneration, revision int) (err error) {
+	k.Lock()
+	defer k.Unlock()
+	k.ensureIntermediateLocked(entryID)
+
+	entry, ok := k.data[entryID.TeamID][entryID.Namespace][entryID.EntryKey]
+	if !ok {
+		// this entry didn't exist in the cache, so there's nothing to check
+		return nil
+	}
+	entryHash := k.hash(ciphertext)
+	if revision < entry.Revision {
+		return KVRevisionCacheError{fmt.Sprintf("cache error: revision decreased from %d to %d", entry.Revision, revision)}
+	}
+	if teamKeyGen < entry.TeamKeyGen {
+		return KVRevisionCacheError{fmt.Sprintf("cache error: team key generation decreased from %d to %d", entry.TeamKeyGen, teamKeyGen)}
+	}
+	if revision == entry.Revision {
+		if teamKeyGen != entry.TeamKeyGen {
+			return KVRevisionCacheError{fmt.Sprintf("cache error: at the same revision (%d) team key gen cannot be different: %d -> %d", revision, entry.TeamKeyGen, teamKeyGen)}
+		}
+		if entryHash != entry.EntryHash {
+			return KVRevisionCacheError{fmt.Sprintf("cache error: at the same revision (%d) hash of entry cannot be different: %s -> %s", revision, entry.EntryHash, entryHash)}
+		}
+	}
+	return nil
+}
+
+func (k *KVRevisionCache) Put(mctx libkb.MetaContext, entryID keybase1.KVEntryID, ciphertext *string, teamKeyGen keybase1.PerTeamKeyGeneration, revision int) (err error) {
```

Comment:
The old PutCheck API seemed a little safer to use. What prompted Put ⮐⮑ Check?

Reply:
the separation ensures that the client doesn't persist anything in its cache that doesn't match the server. i realized in the case where the client thinks an entry is good but the server disagrees, the cache won't know to throwaway that entry and try again.